plugins/bcm57xx/fu-bcm57xx-device.c

```diff
@@ -0,0 +1,767 @@
+/*
+ * Copyright (C) 2018-2020 Evan Lojewski
+ * Copyright (C) 2020 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: GPL-2+
+ */
+
+#include "config.h"
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <errno.h>
+#ifdef HAVE_MMAN_H
+#include <sys/mman.h>
+#endif
+#ifdef HAVE_ETHTOOL_H
+#include <linux/ethtool.h>
+#include <linux/sockios.h>
+#include <net/if.h>
+#endif
+#ifdef HAVE_IOCTL_H
+#include <sys/ioctl.h>
+#endif
+#ifdef HAVE_SOCKET_H
+#include <sys/socket.h>
+#endif
+
+#ifdef HAVE_VALGRIND
+#include <valgrind.h>
+#endif /* HAVE_VALGRIND */
+
+#include "fu-common.h"
+#include "fu-bcm57xx-common.h"
+#include "fu-bcm57xx-device.h"
+#include "fu-bcm57xx-firmware.h"
+
+typedef struct {
+	guint8	*buf;
+	gsize	 bufsz;
+} FuBcm57xxMmap;
+
+#define FU_BCM57XX_BAR_DEVICE	0
+#define FU_BCM57XX_BAR_APE	2
+#define FU_BCM57XX_BAR_MAX	3
+
+struct _FuBcm57xxDevice {
+	FuUdevDevice		 parent_instance;
+	FuBcm57xxMmap		 bar[FU_BCM57XX_BAR_MAX];
+	gchar			*ethtool_iface;
+	int			 ethtool_fd;
+};
+
+G_DEFINE_TYPE (FuBcm57xxDevice, fu_bcm57xx_device, FU_TYPE_UDEV_DEVICE)
+
+static void
+fu_bcm57xx_device_to_string (FuUdevDevice *device, guint idt, GString *str)
+{
+	FuBcm57xxDevice *self = FU_BCM57XX_DEVICE (device);
+	fu_common_string_append_kv (str, idt, "EthtoolIface", self->ethtool_iface);
+}
+
+static void
+fu_bcm57xx_device_ensure_ethtool_iface (FuBcm57xxDevice *self)
+{
+	FuUdevDevice *device = FU_UDEV_DEVICE (self);
+	g_autofree gchar *fn = NULL;
+	g_autoptr(GPtrArray) ifaces = NULL;
+
+	/* do we have a driver (e.g. tg3) providing an ethtool interface */
+	fn = g_build_filename (fu_udev_device_get_sysfs_path (device), "net", NULL);
+	ifaces = fu_common_filename_glob (fn, "en*", NULL);
+	if (ifaces != NULL && ifaces->len > 0) {
+		self->ethtool_iface = g_path_get_basename (g_ptr_array_index (ifaces, 0));
+	} else {
+		g_clear_pointer (&self->ethtool_iface, g_free);
+	}
+}
+
+static gboolean
+fu_bcm57xx_device_probe (FuUdevDevice *device, GError **error)
+{
+	FuBcm57xxDevice *self = FU_BCM57XX_DEVICE (device);
+
+	/* only enumerate number 1 */
+	if (fu_udev_device_get_number (device) != 1) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "only device 1 supported on multi-device card");
+		return FALSE;
+	}
+
+	/* success */
+	fu_bcm57xx_device_ensure_ethtool_iface (self);
+	return fu_udev_device_set_physical_id (device, "pci", error);
+}
+
+#ifdef __ppc64__
+#define BARRIER()	__asm__ volatile ("sync 0\neieio\n" : : : "memory")
+#else
+#define BARRIER()	__asm__ volatile ("" : : : "memory");
+#endif
+
```

Comment:
I believe it should work with lockdown; I don't _think_ I'm writing to writing random system memory, I think the driver is writing to a `mmap`'d PCI resource.

Reply:
Ohh, the barrier is just because we're reading from a mmio resource; I don't think it's hugely important for x64 but for ppc64le it seems like it's more so.

Reply:
Why can't it use `fu_udev_device_pread` or `fu_udev_device_pwrite` then?

Reply:
Well, I'm not reading from a fd, I'm reading an address. Or do you think I can `open()` the `resource4` file and then do `pread`?

Reply:
Isn't the resource4 just pointing to pci address space?  and you're interacting with that pci address space?  I would think an ioctl out to that address space should work just fine.